File: sol/341.cpp
#include <stack>

class NestedIterator {
private:
    stack<NestedInteger> content;
public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        for (vector<NestedInteger>::const_reverse_iterator it = nestedList.crbegin();
             it != nestedList.crend();
             ++it){
            this->content.push(*it);
        }
    }
    
    int next() {
        this->hasNext();
        NestedInteger ni = this->content.top();
        this->content.pop();
        return ni.getInteger();
    }
    
    bool hasNext() {
        while (!this->content.empty() && !this->content.top().isInteger()){
            vector<NestedInteger> data = this->content.top().getList();
            this->content.pop();
            for (vector<NestedInteger>::const_reverse_iterator it = data.crbegin();
                 it != data.crend();
                 ++it){
                this->content.push(*it);
            }
        }
        return !this->content.empty();
    }
};

File: sol/341/341.py
class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.stack = nestedList[::-1]

    def next(self) -> int:
        self.hasNext()
        return self.stack.pop().getInteger()

    def hasNext(self) -> bool:
        while len(self.stack) != 0 and not self.stack[-1].isInteger():
            ni = self.stack.pop()
            self.stack.extend(ni.getList()[::-1])
        return len(self.stack) != 0

# class NestedIterator:
#     class NestedIntegerYielder:
#         def __init__(self, data: NestedInteger):
#             self.data = data
# 
#         def __iter__(self) -> Iterator[int]:
#             if self.data.isInteger():
#                 yield self.data.getInteger()
#             else:
#                 for it in (self.__class__(content)
#                         for content in self.data.getList()):
#                     yield from it
# 
#     def __init__(self, nestedList: [NestedInteger]):
#         self.nestedList: list[NestedInteger] = nestedList
#         self.buffered: int | None = None
#         self.yielder: Generator[int] = (y
#             for content in self.nestedList
#             for y in self.NestedIntegerYielder(content))
# 
#     def next(self) -> int:
#         self.hasNext()
#         if self.buffered is not None:
#             temp = self.buffered
#             self.buffered = None
#             return temp
# 
#     def hasNext(self) -> bool:
#         if self.buffered is None:
#             self.buffered = next(self.yielder, None)
#         return self.buffered is not None

